designs/carbon-aware.md

```diff
@@ -0,0 +1,197 @@
+# Carbon Aware Karpenter: Optimizing Kubernetes Cluster Autoscaling for Environmental Sustainability
+*Author: [@JacobValdemar](https://github.com/JacobValdemar)*
+
+## Context & Problem
+There is a growing concern about the environmental impact of Kubernetes clusters. Karpenter's opportunities within environmental sustainability is referenced in multiple comments that back [`karpenter-core`'s move to CNCF](https://github.com/kubernetes/org/issues/4258).
+
+I am currently working on my master's thesis in Computer Engineering (Master of Science in Engineering) at Aarhus University located in Denmark. The objective of the thesis is to enable Karpenter to minimize carbon emissions from Kubernetes clusters that run on cloud infrastructure (scoped to AWS).
+
+RFC: https://github.com/aws/karpenter/issues/4630
+
+## Fundamentals of Green Software
+I will try to keep it simple. The reader should be familiar with the following.
+
+A cluster's emissions is made of two elements: embodied emissions and operational emissions. To get the total emissions, one can add them togeather.
+
+- **Embodied carbon emissions**: Manufacturing emissions (CO₂e) amortized over instance lifetime (usually 4 years) divided by how long we use the instance
+- **Operational carbon emissions**: Carbon emitted by electricity grid to produce electricity for the instance in the region where it is used, multiplied by PUE
+
+There is a lot more to Green Software. If you want to learn more, I recommend you to visit [Green Software Practitioner](https://learn.greensoftware.foundation/) (a Green Software Foundation project - an affiliate of the Linux Foundation).
+
+## Solution
+
+### Feature Gate
+The feature is proposed to be controlled using a [feature gate](https://karpenter.sh/docs/concepts/settings/#feature-gates).
+
+| **Feature** | **Default** |         **Config Key**          | **Stage** |    **Since**    | **Until** |
+| :---------: | :---------: | :-----------------------------: | :-------: | :-------------: | :-------: |
+| CarbonAware |    false    | featureGates.carbonAwareEnabled |   Alpha   | v0.32.0/v0.33.0 |           |
+
+### Carbon emissions data source
+Currently the best option is to create estimates based on the methodology used in [Boaviztapi](https://github.com/Boavizta/boaviztapi).
+
+[Try out Boaviztapi on the Datavizta demo website](https://datavizta.boavizta.org/cloudimpact).
+
+#### Licensing
+Boaviztapi is licensed under [`GNU Affero General Public License v3.0`](https://github.com/Boavizta/boaviztapi/blob/main/LICENSE). Therefore, as far as I know, we must license their data under the same license if used in the Karpenter repository.
```

Comment:
Will there be conflicts in licensing their data according to these requirements and the CNCF guidance for licensing the karpenter-core source itself?

Reply:
@jackfrancis So I am in no way a lawyer, but I guess that there wont be licensing issues. We probably just have to put a different license notice in the top of files that is based on their works. Currently, I would expect the data with that license only to be in cloud provider repos (e.g. https://github.com/aws/karpenter) and not karpenter-core.

Reply:
> So I am in no way a lawyer, but I guess that there wont be licensing issues

Famous last words.

Reply:
We'd definitely need to check with AWS legal for this (while AWS owns it) and consider how this interacts with the CNCF guidance since this would probably live in the karpenter-core repo if it was gen-ed. I could also see this living in its own separate repo that provided configuration plugins to the pricing overrides that we are thinking about here

Reply:
The particular license that Boaviztapi chooses to use suggests that it's using the "copyleft malware" approach (for Good, not Evil, of course) to ensure that its definitions of freedom (as in both beer and speech) are enforced upon downstream projects. So that's the only reason to bring this up. AWS and CNCF are probably going to combine w/ such a license in their own unique interesting ways. Non-zero chance there will be friction.